generic-contracts/src/main/java/com/scalar/dl/genericcontracts/table/v1_0_0/Select.java

```diff
@@ -53,18 +160,249 @@ private JsonNode project(JsonNode records, JsonNode projections) {
+        || !arguments.has(Constants.QUERY_TABLE)
+        || !arguments.has(Constants.QUERY_CONDITIONS)) {
+      throw new ContractContextException(Constants.INVALID_QUERY_FORMAT);
+    }
+
+    // Check the first table
+    validateTable(arguments.get(Constants.QUERY_TABLE));
+    Set<String> tablesReferences = new HashSet<>();
+    String firstTableReference = getTableReference(arguments.get(Constants.QUERY_TABLE));
+    tablesReferences.add(firstTableReference);
+
+    // Check joins
+    if (arguments.has(Constants.QUERY_JOINS)) {
+      validateJoins(firstTableReference, arguments.get(Constants.QUERY_JOINS));
+      tablesReferences.addAll(getJoinTableReferences(arguments.get(Constants.QUERY_JOINS)));
+    }
+
+    // Check conditions
+    validateConditions(arguments.get(Constants.QUERY_CONDITIONS), tablesReferences);
+
+    // Check projections
+    if (arguments.has(Constants.QUERY_PROJECTIONS)) {
+      validateProjections(arguments.get(Constants.QUERY_PROJECTIONS), tablesReferences);
+    }
+  }
+
+  private void validateColumn(String column, Set<String> tableReferences) {
+    boolean hasTableReference = isColumnNameWithTableReference(column);
+
+    if (!hasTableReference) {
+      // Join queries must have a table reference
+      if ((tableReferences.size() == 1 && !isColumnName(column)) || tableReferences.size() > 1) {
+        throw new ContractContextException(Constants.INVALID_COLUMN_FORMAT + column);
+      }
+      return;
+    }
+
+    // Check if the table of the column is appeared in the query
+    String tableReference = getTableReference(column);
+    if (!tableReferences.contains(tableReference)) {
+      throw new ContractContextException(Constants.UNKNOWN_TABLE + tableReference);
+    }
+  }
+
+  private void validateTable(JsonNode table) {
+    if ((!table.isTextual() && !table.isObject())
+        || (table.isObject()
+            && (table.size() != 2
+                || !table.has(Constants.ALIAS_NAME)
+                || !table.get(Constants.ALIAS_NAME).isTextual()
+                || !table.has(Constants.ALIAS_AS)
+                || !table.get(Constants.ALIAS_AS).isTextual()))) {
+      throw new ContractContextException(Constants.INVALID_QUERY_TABLE_FORMAT + table);
+    }
+  }
+
+  private void validateJoins(String firstTableReference, JsonNode joins) {
+    Set<String> seenTables = new HashSet<>();
+    seenTables.add(firstTableReference);
+
+    if (!joins.isArray()) {
+      throw new ContractContextException(Constants.INVALID_QUERY_FORMAT);
+    }
+
+    for (JsonNode join : joins) {
+      // Check the join object format
+      if (!join.isObject()
+          || join.size() != 3
+          || !join.has(Constants.JOIN_TABLE)
+          || !join.has(Constants.JOIN_LEFT_KEY)
+          || !join.has(Constants.JOIN_RIGHT_KEY)) {
+        throw new ContractContextException(Constants.INVALID_JOIN_FORMAT + join);
+      }
+
+      // Check the join table format
```

Comment:
I've changed the specification for the join keys based on the discussion in the demo meeting since the previous format is somewhat ambiguous. You can use `left` and `right` for specifying the left and right join keys, respectively. `left_key(s)` and `left_column(s)` were also the candidates, but I would like to keep them not only short but also compatible with both singular and multiple keys since we can only use a single column for the join key for now.